CRM/Contact/BAO/Relationship.php

```diff
@@ -2404,21 +2407,78 @@ private static function isRelationshipTypeCurrentEmployer(int $existingTypeID):
+    if (!$membershipInheritedFrom || !in_array($values[$cid]['relationshipTypeId'], $relTypeIds)) {
       return [implode(',', $relTypeIds), FALSE];
     }
     //CRM-16300 check if owner membership exist for related membership
-    $isDeletable = !empty($values[$mainRelatedContactId]['memberships'][$membshipInheritedFrom]);
-    return [implode(',', $relTypeIds), $isDeletable];
+    return [implode(',', $relTypeIds), self::isContactHasValidRelationshipToInheritMembershipType((int) $cid, (int) $membershipValues['membership_type_id'], (int) $membershipValues['owner_membership_id'])];
+  }
+
+  /**
+   * Is there a valid relationship confering this membership type on this contact.
+   *
+   * @param int $contactID
+   * @param int $membershipTypeID
+   * @param int $parentMembershipID
+   *   Id of the membership being inherited.
+   *
+   * @return bool
+   *
+   * @throws \CiviCRM_API3_Exception
+   */
+  private static function isContactHasValidRelationshipToInheritMembershipType(int $contactID, int $membershipTypeID, int $parentMembershipID): bool {
+    $membershipType = CRM_Member_BAO_MembershipType::getMembershipType($membershipTypeID);
+    $existingRelationships = civicrm_api3('Relationship', 'get', [
+      'contact_id_a' => $contactID,
+      'contact_id_b' => $contactID,
+      'relationship_type_id' => ['IN' => $membershipType['relationship_type_id']],
+      'options' => ['or' => [['contact_id_a', 'contact_id_b']], 'limit' => 0],
+      'is_active' => 1,
+    ])['values'];
+
+    if (empty($existingRelationships)) {
+      return FALSE;
+    }
+
+    $membershipInheritedFromContactID = civicrm_api3('Membership', 'getvalue', ['return' => 'contact_id', 'id' => $parentMembershipID]);
+    // I don't think the api can correctly filter by start & end because of handling for NULL
+    // so we filter them out here.
+    foreach ($existingRelationships as $index => $existingRelationship) {
+      if (!empty($existingRelationship['start_date'])
+        && strtotime($existingRelationship['start_date']) > time()
+      ) {
+        unset($existingRelationships[$index]);
+        continue;
+      }
+      if (!empty($existingRelationship['end_date'])
+        && strtotime($existingRelationship['end_date']) < time()
+      ) {
+        unset($existingRelationships[$index]);
+        continue;
+      }
+      if ($membershipInheritedFromContactID !== (int) $existingRelationship['contact_id_a']
+        && $membershipInheritedFromContactID !== (int) $existingRelationship['contact_id_b']
+      ) {
+        // This is a weird scenario - they would still be entitled to have the membership
+        // just not from this relationship - and some max_related calcs etc would be required.
+        // For now ignore here & hope it's handled elsewhere - at least that's consistent with
+        // before this function was added.
+        unset($existingRelationships[$index]);
+        continue;
+      }
```

Comment:
Not sure if this should be checked or not. I think max_related is not needed here since the membership is not deleted and re-created.
Test failed because of here.

Reply:
Just cannot think of a case that needs this check.
@eileenmcnaughton any hints?